milli/src/search/new/matches/matching_words.rs

```diff
@@ -0,0 +1,323 @@
+use std::cmp::Reverse;
+use std::fmt;
+use std::ops::RangeInclusive;
+
+use charabia::Token;
+
+use super::super::interner::Interned;
+use super::super::query_term::LocatedQueryTerm;
+use super::super::{DedupInterner, Phrase};
+use crate::SearchContext;
+
+pub struct LocatedMatchingPhrase {
+    pub value: Interned<Phrase>,
+    pub positions: RangeInclusive<WordId>,
+}
+
+pub struct LocatedMatchingWords {
+    pub value: Vec<Interned<String>>,
+    pub positions: RangeInclusive<WordId>,
+    pub is_prefix: bool,
```

Comment:
(Not so important) question @ManyTheFish : what do we need the positions for, could they be replaced with a generic `id` that identifies a word/term/phrase?

Reply:
Yes but the condition is that the id should refer to a term that was originally written in the query, for instance a word and all there synonyms must share the same `id` for matching_words